File: tests/mocks/mockDockerEngine.ts
import { vi } from "vitest";

export class MockDockerEngine {
  containers: Array<{
    Id: string;
    Names: string[];
    State: string;
    Labels: Record<string, string>;
    Config?: { Env?: string[]; Cmd?: string[]; Image?: string };
  }> = [];
  localImages = new Map<
    string,
    {
      Id: string;
      RepoTags: string[];
      Size: number;
      Architecture: string;
      Os: string;
      Created: string;
    } | null
  >();
  pullImageLines: string[] = [];
  listContainers = vi.fn(
    async (_opts: { all?: boolean; filters?: unknown } = {}) => this.containers,
  );
  inspect = vi.fn(async (id: string) => this.containers.find((c) => c.Id === id) ?? {});
  getContainer = vi.fn((id: string) => ({
    inspect: vi.fn(async () => this.containers.find((c) => c.Id === id) ?? {}),
  }));
  inspectImage = vi.fn(async (nameOrId: string) => {
    if (this.localImages.has(nameOrId)) return this.localImages.get(nameOrId) ?? null;
    return {
      Id: `sha256:${nameOrId}`,
      RepoTags: [nameOrId],
      Size: 1,
      Architecture: "amd64",
      Os: "linux",
      Created: "2026-01-01T00:00:00.000Z",
    };
  });
  listImages = vi.fn(async () =>
    [...this.localImages.values()].filter(
      (
        image,
      ): image is {
        Id: string;
        RepoTags: string[];
        Size: number;
        Architecture: string;
        Os: string;
        Created: string;
      } => image !== null,
    ),
  );
  pullImage = vi.fn(
    async function* (
      this: MockDockerEngine,
      _image: string,
      _opts: { signal?: AbortSignal } = {},
    ): AsyncGenerator<string, void> {
      for (const line of this.pullImageLines) yield line;
    }.bind(this),
  );
}

File: tests/mocks/mock_docker_engine.py
"""Mock Docker engine for tool tests."""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any

from docker_agent.services.docker.types import (
    ContainerInspect,
    ContainerStats,
    ContainerSummary,
    ImageInspect,
    ImageSummary,
)


class MockDockerEngine:
    def __init__(self) -> None:
        self.containers: list[dict[str, Any]] = []
        self.local_images: dict[str, ImageInspect | None] = {}
        self.pull_image_lines: list[str] = []
        self.stats_by_id: dict[str, Any] = {}
        self.inspect_by_id: dict[str, Any] = {}
        self.list_containers_error: BaseException | None = None
        self.pull_image_calls: list[tuple[str, dict[str, Any]]] = []

    async def list_containers(
        self, *, all: bool = False, filters: dict[str, list[str]] | None = None
    ) -> list[ContainerSummary]:
        del all, filters
        if self.list_containers_error is not None:
            raise self.list_containers_error
        return [
            ContainerSummary.model_validate(container)
            for container in self.containers
        ]

    async def inspect(self, container_id: str) -> ContainerInspect:
        explicit = self.inspect_by_id.get(container_id)
        if explicit is not None:
            return ContainerInspect.model_validate(explicit)
        for container in self.containers:
            if container.get("Id") == container_id:
                return ContainerInspect.model_validate(container)
        return ContainerInspect.model_validate({"Id": container_id})

    async def stats(self, container_id: str) -> ContainerStats:
        return ContainerStats.model_validate(
            self.stats_by_id.get(container_id, {})
        )

    async def inspect_image(self, name_or_id: str) -> ImageInspect | None:
        if name_or_id in self.local_images:
            return self.local_images[name_or_id]
        return ImageInspect.model_validate(
            {
                "Id": f"sha256:{name_or_id}",
                "RepoTags": [name_or_id],
                "Size": 1,
                "Architecture": "amd64",
                "Os": "linux",
                "Created": "2026-01-01T00:00:00.000Z",
            }
        )

    async def list_images(
        self, *, filters: dict[str, list[str]] | None = None
    ) -> list[ImageSummary]:
        del filters
        return [
            ImageSummary.model_validate(
                {
                    "Id": image.Id,
                    "RepoTags": image.repo_tags,
                    "Size": image.size,
                    "Created": 0,
                }
            )
            for image in self.local_images.values()
            if image is not None
        ]

    async def pull_image(
        self, image: str, *, signal: Any | None = None
    ) -> AsyncIterator[str]:
        self.pull_image_calls.append((image, {"signal": signal}))
        for line in self.pull_image_lines:
            yield line